semgrep-core/matching/Rules_filter.ml

```diff
@@ -0,0 +1,155 @@
+ *)
+open Common
+module Flag = Flag_semgrep
+module R = Rule
+module V = Visitor_AST
+open AST_generic
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* Rules filtering using regexps.
+ *
+ * It is useless to run the engine on rules containing specific IDs
+ * (e.g., eval) that are never mentioned in the file.
+ *
+ * We did something similar in Coccinelle I think. This also has been
+ * mentioned many times (by Clint, HN, etc.).
+ *
+ * notes: I tried to use the ocaml-re (Re) regexp libray instead of Str
+ * because I thought it would be faster, and because it offers regexp
+ * combinators (alt, rep, etc.) which might be useful at some point to
+ * handle patterns containing explicit DisjExpr. However when running
+ * on Zulip codebase with zulip semgrep rules, Str is actually faster
+ * than Re.
+ *
+ * todo:
+ *  - we could even avoid parsing the file by using a lazy AST
+ *    in Semgrep_generic.check.
+ *)
+
+(*****************************************************************************)
+(* Helpers  *)
+(*****************************************************************************)
+
+(* Those functions using Re instead of Str are actually slower on
+ * Zulip.
+let regexp_matching_str s =
+  Re.str s
+let compile_regexp t =
+  Common.profile_code "Re.compile_regexp" (fun () -> Re.compile t)
+let run_regexp re str =
+  Common.profile_code "Re.run_regexp" (fun () -> Re.execp re str)
+*)
+
+let regexp_matching_str s =
+  Str.regexp_string s
+let compile_regexp t =
+  Common.profile_code "Re.compile_regexp" (fun () -> t)
+let run_regexp re str =
+  Common.profile_code "Re.run_regexp" (fun () ->
+   (* bugfix:
+    * this does not work!:  Str.string_match re str 0
+    * because you need to add ".*" in front to make it work,
+    * (but then you can not use regexp_string above)
+    * => use Str.search_forward instead.
+    *)
+   try
+     Str.search_forward re str 0 |> ignore; true
+   with Not_found -> false
+  )
+
+let reserved_id lang str =
+  Metavars_generic.is_metavar_name str ||
+  (* in JS field names can be regexps *)
+  (lang = Lang.Javascript && Matching_generic.is_regexp_string str) ||
+  (* ugly hack that we then need to handle also here *)
+  str = AST_generic.special_multivardef_pattern ||
+  (* ugly: because ast_js_build introduce some extra "!default" ids *)
+  (lang = Lang.Javascript && str = Ast_js.default_entity) ||
+  (* parser_js.mly inserts some implicit this *)
+  (lang = Lang.Java && str = "this")
+
+let reserved_str str =
+  str = "..." ||
+  Matching_generic.is_regexp_string str
+
+(*****************************************************************************)
+(* ID extractor  *)
+(*****************************************************************************)
+let extract_specific_strings lang any =
+  let res = ref [] in
+  let visitor = V.mk_visitor {V.default_visitor with
```

Comment:
Is this correctly exhaustive?

I'm wondering what will happen with file and package names, e.g.:

```go
import "fmt"
```

```java
package foo;
```

Reply:
Reflecting on ^^, I think we need a mechanism to log which rules are rejected for each file. Can we print this to stderr in verbose / debug mode?

Reply:
Yes use -debug

Reply:
It was not exhaustive first, but after a few iterations i got make test and semgreprule test working